### PR TITLE
Implement `SAVE` attribute

### LINKED ISSUE #556: Add save attribute and implicit save for arrays
*Original issue*: https://gitlab.com/lfortran/lfortran/-/issues/721

Variable saving in functions and subroutines needs to be added for arrays. The idea of local variables allocated in global variables is discussed in issue https://gitlab.com/lfortran/lfortran/-/issues/713. Here is a test case to work toward satisfying.
```
program save_with_arrays
real :: x
x = func1()
print *, "4.0 == ", x
x = func1()
print *, "6.0 == ", x

contains

function func1() result(a)
    real :: saved(2) = [1.0, 1.0]
    real :: a
    saved = saved + 1.0
    a = sum(saved)
end function
end program save_with_arrays
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #556.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -1458,8 +1458,17 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                                    Iterate over all variables in the symbol table
+                                    and set the storageType to Save
+                                */
+                                for (auto &it: current_scope->get_scope()) {
+                                    ASR::symbol_t* sym = it.second;
+                                    if (ASR::is_a<ASR::Variable_t>(*sym)) {
+                                        ASR::Variable_t* var = ASR::down_cast<ASR::Variable_t>(sym);
+                                        var->m_storage = ASR::storage_typeType::Save;
+                                    }
+                                }
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1471]: I think you also need to set "default storage" variable to "true", and then set all subsequent declaration as "save" as well.

File: integration_tests/save_02.f90 [hunk 2]
```diff
@@ -0,0 +1,7 @@
+program save_02
+implicit none
+integer :: i, j
+real :: k, l
+double precision :: m(5), n(5, 5), o(5, 6, 7)
+save
+end program
```
Comment (certik) [inline integration_tests/save_02.f90:7]: Also add a test:
```
integer :: i, j
save
real :: k, l
double precision :: m(5), n(5, 5), o(5, 6, 7)
```

**Review by certik — APPROVED** [review]
I think this is good, but there are more cases that need fixing. You can either do it as part of this PR or subsequent PRs.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1471]: I am sorry but I did not understand what you are trying to tell here.

Comment (certik) [reply src/lfortran/semantics/ast_common_visitor.h:1471]: I was suggesting how to update it to make it work for this case:
```
integer :: i, j
save
real :: k, l
double precision :: m(5), n(5, 5), o(5, 6, 7)
```
When you encounter "save", you want to go over all already declared variables and mark them save, as you already do. But then you also want to do something like `default_storage = save` (where `default_storage` is a visitor variable), and then later, when you encounter all other declarations, like `real :: k, l`, set the storage using "default_storage". By default, "default_storage" will be initialized as "not save". Let me know if it is more clear now.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1471]: Ohh okay, `save` present here must be applied to 
```
real :: k, l
double precision :: m(5), n(5, 5), o(5, 6, 7)
```
these as well. I thought we must save only `integer :: i, j`.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1471]: Perfect. Thank you for the clarification. I will implement that by tomorrow :)

Comment (certik) [reply src/lfortran/semantics/ast_common_visitor.h:1471]: Yes, `save` applies to all the variables in the function, no matter where you put it.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -773,6 +773,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
     ASR::ttype_t* current_variable_type_;
 
     int32_t enum_init_val;
+    int default_storage_save = 0;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:776]: I am worried this if fragile to be increasing an int.

Let's do:
```
bool default_storage_save
```
Then set it to false when entering a subroutine/function/program/module.

When you encounter "save", set it to true.

**Comment by certik** [comment]
Let's also add tests for:
```
   integer :: i, j
   save
   real :: k, l
```
Inside a function and subroutine and module. Right now it is only Program.

**Comment by certik** [comment]
Other than that, I think this is good, thanks!

File: src/lfortran/semantics/ast_common_visitor.h (hunk 3 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:776]: I initially did that, what about cases of nested subroutines?

Comment (certik) [reply src/lfortran/semantics/ast_common_visitor.h:776]: You need to reset it before visiting the nested subroutines. Let's add a test for that too, to ensure it works.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:776]: Yes got it. I was confused that we can declarations after `contains` as well but actually we can't. Thank you for clarification.

**Review by certik — APPROVED** [review]
I think this is good, thanks!

**Pull request merged**